Game.Checkers/CheckersState.cs

```diff
@@ -1,10 +1,159 @@
+        //private static Dictionary<(int, int), (int, int)[] Neighbours = new Dictionary<(int, int), List<(int, int)>> {
+        //    new KeyValuePair<(int, int),List<(int, int)>>((0,0), new List<(int, int)> { (1,1) }),
+        //};
+        private Piece[,] board;
+        public Player CurrentPlayer { get; private set; }
+        private CheckersState(Piece[,] board, Player currentPlayer)
+        {
+            this.board = new Piece[BOARD_COLS, BOARD_ROWS];
+            Array.Copy(board, this.board, board.Length);
+            CurrentPlayer = currentPlayer;
+        }
+        public CheckersState(CheckersState state)
+        {
+            board = new Piece[BOARD_COLS, BOARD_ROWS];
+            Array.Copy(state.board, board, state.board.Length);
+            CurrentPlayer = state.CurrentPlayer;
+        }
         public bool Equals(CheckersState? other)
         {
-            throw new NotImplementedException();
+            if (other == null) return false;
+            if (CurrentPlayer != other.CurrentPlayer) return false;
+            for (int x = 0; x < BOARD_ROWS; x++)
+            {
+                for (int y = 0; y < BOARD_COLS; y++)
+                {
+                    if (GetPieceAt(x, y) != other.GetPieceAt(x, y)) return false;
+                }
+            }
+            return true;
+        }
+        public IEnumerable<Field> GetFields()
+        {
+            for (int col = 0; col < BOARD_ROWS; col++)
+            {
+                for (int row = col % 2; row < BOARD_COLS; row += 2)
+                {
+                    yield return new Field(col, row);
+                }
+            }
+        }
+        public IEnumerable<Field> GetNeighbours(Field field)
+        {
+            int x = field.Col;
+            int y = field.Row;
+            List<Field> neighbours = new();
+            if (y < BOARD_ROWS - 1)
+            {
+                if (x < BOARD_COLS - 1)
+                    neighbours.Add(new Field(x + 1, y + 1));
+                if (x > 0)
+                    neighbours.Add(new Field(x - 1, y + 1));
+            }
+            if (y > 0)
+            {
+                if (x < BOARD_COLS - 1)
+                    neighbours.Add(new Field(x + 1, y - 1));
+                if (x > 0)
+                    neighbours.Add(new Field(x - 1, y - 1));
+            }
+            return neighbours;
+        }
+
+        public Piece GetPieceAt(int col, int row)
+        {
+            return board[col, row];
+        }
+
+        public Piece GetPieceAt(Field field)
+        {
+            return GetPieceAt(field.Col, field.Row);
+        }
+
+        public void SetPieceAt(int col, int row, Piece piece)
+        {
+            board[col, row] = piece;
+        }
+        public void SetPieceAt(Field field, Piece piece)
+        {
+            SetPieceAt(field.Col, field.Row, piece);
+        }
+        public void SetPieceAt(Field field, Piece piece, bool possiblePromotion)
```

Comment:
czy possiblePromotion i sprawdzenie czy jest w ostatnim rzędzie to nie jest powtórzenie tego samego?

Reply:
To kwestia tego, że jak mamy wielokrotne bicie pionem i przechodzimy przez ostatni rząd to nie awansujemy na damkę. Możliwość awansu sprawdzamy tylko jeśli kończymy ruch w ostatnim rzędzie.